server/api/auth/createWithIdp.js

```diff
@@ -0,0 +1,82 @@
+const http = require('http');
+const https = require('https');
+const sharetribeSdk = require('sharetribe-flex-sdk');
+const { handleError, serialize, typeHandlers } = require('../../api-util/sdk');
+
+const CLIENT_ID = process.env.REACT_APP_SHARETRIBE_SDK_CLIENT_ID;
+const CLIENT_SECRET = process.env.SHARETRIBE_SDK_CLIENT_SECRET;
+const TRANSIT_VERBOSE = process.env.REACT_APP_SHARETRIBE_SDK_TRANSIT_VERBOSE === 'true';
+const USING_SSL = process.env.REACT_APP_SHARETRIBE_USING_SSL === 'true';
+const BASE_URL = process.env.REACT_APP_SHARETRIBE_SDK_BASE_URL;
+
+const FACBOOK_APP_ID = process.env.REACT_APP_FACEBOOK_APP_ID;
+const GOOGLE_CLIENT_ID = process.env.REACT_APP_GOOGLE_CLIENT_ID;
+
+// Instantiate HTTP(S) Agents with keepAlive set to true.
+// This will reduce the request time for consecutive requests by
+// reusing the existing TCP connection, thus eliminating the time used
+// for setting up new TCP connections.
+const httpAgent = new http.Agent({ keepAlive: true });
+const httpsAgent = new https.Agent({ keepAlive: true });
+
+const baseUrl = BASE_URL ? { baseUrl: BASE_URL } : {};
+
+module.exports = (req, res) => {
+  const tokenStore = sharetribeSdk.tokenStore.expressCookieStore({
+    clientId: CLIENT_ID,
+    req,
+    res,
+    secure: USING_SSL,
+  });
+
+  const sdk = sharetribeSdk.createInstance({
+    transitVerbose: TRANSIT_VERBOSE,
+    clientId: CLIENT_ID,
+    clientSecret: CLIENT_SECRET,
+    httpAgent,
+    httpsAgent,
+    tokenStore,
+    typeHandlers,
+    ...baseUrl,
+  });
+
+  const { idpToken, source, ...rest } = req.body;
+
+  // Choose the idpClientId based on which authentication method is used.
+  // Currently available options are "Facebook" and "Google"
+  const idpClientId =
+    source === 'Facebook' ? FACBOOK_APP_ID : source === 'Google' ? GOOGLE_CLIENT_ID : null;
+
+  sdk.currentUser
+    .createWithIdp({
+      idpId: idpClientId,
+      idpToken,
+      ...rest,
+    })
+    .then(() =>
+      // After the user is created, we need to call authWithIdp endpoint
+      // so that the user will be logged in.
+      sdk.authWithIdp({
+        idpClientId,
+        idpToken,
+      })
+    )
+    .then(apiResponse => {
+      const { status, statusText, data } = apiResponse;
+      res
+        .clearCookie('authinfo')
+        .status(status)
+        .set('Content-Type', 'application/transit+json')
+        .send(
+          serialize({
+            status,
+            statusText,
+            data,
+          })
+        )
+        .end();
+    })
+    .catch(e => {
+      handleError(res, e);
```

Comment:
What happens to cookie?
(I assume that it should be cleared here (and ensure it also on logout)

Reply:
The cookie will be cleared if the authentication is successful or it will expire after 15 min (or this time can be changed). 

I'm not sure if the cookie should be cleared here when there is an error because then the user is not able to try the confirm step again. E.g. the error could be a conflicting email and then the user could try again with a different email (but I'm not sure how likely something like this would happen. One scenario would be that user has signed up with the same email as in e.g. Facebook but not verified the email address yet and for some reason they would decide to use different email in Flex).